packages/api-server/rest_server/tasks/dispatcher.py

```diff
@@ -0,0 +1,216 @@
+        task_profile format. add this accordingly when a new msg field
+        is introduced.
+        The 'start time' here is refered to the "Duration" from now.
+        """
+        req_msg = SubmitTask.Request()
+
+        try:
+            if (
+                ("task_type" not in task_json)
+                or ("start_time" not in task_json)
+                or ("description" not in task_json)
+            ):
+                raise Exception("Key value is incomplete")
+
+            if "priority" in task_json:
+                priority = int(task_json["priority"])
+                if priority < 0:
+                    raise Exception("Priority value is less than 0")
+                req_msg.description.priority.value = priority
+            else:
+                req_msg.description.priority.value = 0
+
+            desc = task_json["description"]
+            if task_json["task_type"] == "Clean":
+                req_msg.description.task_type.type = TaskType.TYPE_CLEAN
+                req_msg.description.clean.start_waypoint = desc["cleaning_zone"]
+            elif task_json["task_type"] == "Loop":
+                req_msg.description.task_type.type = TaskType.TYPE_LOOP
+                loop = Loop()
+                loop.num_loops = int(desc["num_loops"])
+                loop.start_name = desc["start_name"]
+                loop.finish_name = desc["finish_name"]
+                req_msg.description.loop = loop
+            elif task_json["task_type"] == "Delivery":
+                req_msg.description.task_type.type = TaskType.TYPE_DELIVERY
+                delivery = Delivery()
+                delivery.pickup_place_name = desc["pickup_place_name"]
+                delivery.pickup_dispenser = desc["pickup_dispenser"]
+                delivery.dropoff_ingestor = desc["dropoff_ingestor"]
+                delivery.dropoff_place_name = desc["dropoff_place_name"]
+                req_msg.description.delivery = delivery
+            else:
+                raise Exception("Invalid TaskType")
+
+            # Calc start time, convert min to sec: TODO better representation
+            rclpy.spin_once(self, timeout_sec=0.0)
+            ros_start_time = self.get_clock().now().to_msg()
+            ros_start_time.sec += int(task_json["start_time"] * 60)
+            req_msg.description.start_time = ros_start_time
+        except KeyError as ex:
+            return None, f"Missing Key value in json body: {ex}"
+        except Exception as ex:
+            return None, str(ex)
+        return req_msg, ""
+
+    def get_task_status(self):
+        """
+        Get all task status - This fn will trigger a ros srv call to acquire
+        all submitted tasks to dispatcher node. Fn returns an object of tasks
+        """
+        req = GetTaskList.Request()
+        try:
+            future = self.get_tasks_srv.call_async(req)
+            rclpy.spin_until_future_complete(self, future, timeout_sec=0.5)
```

Comment:
Some things to test:

Because we will need to publish stuffs, test if `publish` can work if rclpy is not spinning, or if there is a way to publish with `spin_until_future_complete`. If that didn't work out, then we can't use this pattern. We will need to start a new thread and do `rclpy.spin` there, then we need some way to synchronize the event loop. We could use `call()` here, but that is a blocking call and would destroy performance.

In any case, `spin_until_future_complete` is blocking and would destroy perf so we should probably move away from this pattern.

Reply:
So there would be 2 threads, one to spin for the result of the service (task details) and the other to spin the publisher? The publisher is going to publish other things but if a task is returned, then we would need to synchronize the 2 threads to publish a `task event` or something like that?

Pardon me for my ignorance, but what would we be publishing? I thought the `rest server` is just to handle a one time request and the change in robot states and trajectories would be handled by the `api-server`, where we can subscribe to it from the dashboard?

Reply:
There would be 1 spinning thread that takes care of all rclpy operations. Door and lift requests are an example where we need to publish stuffs.

I'm working on the infrastructure apis, along with some refactoring, so I guess you can just ignore this, let's merge this first, then I will refactor this in my PR.